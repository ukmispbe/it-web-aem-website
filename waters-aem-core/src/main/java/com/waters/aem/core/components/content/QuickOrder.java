package com.waters.aem.core.components.content;

import com.adobe.cq.export.json.ComponentExporter;
import com.adobe.cq.export.json.ExporterConstants;
import com.citytechinc.cq.component.annotations.Component;
import com.citytechinc.cq.component.annotations.DialogField;
import com.citytechinc.cq.component.annotations.Listener;
import com.citytechinc.cq.component.annotations.widgets.PathField;
import com.citytechinc.cq.component.annotations.widgets.TextField;
import com.icfolson.aem.library.api.link.Link;
import com.icfolson.aem.library.models.annotations.LinkInject;
import com.waters.aem.core.components.SiteContext;
import com.waters.aem.core.constants.WatersConstants;
import com.waters.aem.core.services.commerce.WatersCommerceService;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.models.annotations.DefaultInjectionStrategy;
import org.apache.sling.models.annotations.Exporter;
import org.apache.sling.models.annotations.Model;
import org.apache.sling.models.annotations.injectorspecific.OSGiService;
import org.apache.sling.models.annotations.injectorspecific.Self;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import static com.icfolson.aem.library.core.constants.ComponentConstants.*;

@Component(value = "Quick Order",
        description = "This is the Quick Order component for Waters site",
        listeners = {
                @Listener(name = EVENT_AFTER_EDIT, value = REFRESH_PAGE),
                @Listener(name = EVENT_AFTER_MOVE, value = REFRESH_PAGE),
                @Listener(name = EVENT_AFTER_COPY, value = REFRESH_PAGE)
        })
@Model(adaptables = SlingHttpServletRequest.class,
        adapters = {QuickOrder.class, ComponentExporter.class},
        resourceType = QuickOrder.RESOURCE_TYPE,
        defaultInjectionStrategy = DefaultInjectionStrategy.OPTIONAL)
@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME,
        extensions = ExporterConstants.SLING_MODEL_EXTENSION)
public class QuickOrder implements ComponentExporter {

    public static final String RESOURCE_TYPE = "waters/components/content/quickorder";

    @Self
    private SiteContext siteContext;

    @OSGiService
    private WatersCommerceService watersCommerceService;

    @DialogField(fieldLabel = "Button Label",
            fieldDescription = "The text to display on the Button",
            defaultValue = "ADD TO CART",
            ranking = 1)
    @TextField
    @Inject
    private String buttonLabel;

    public String getButtonLabel() {
        return buttonLabel;
    }

    @DialogField(fieldLabel = "Add to Cart Placeholder",
            fieldDescription = "The text to display as placeholder",
            defaultValue = "SKU Number",
            ranking = 2)
    @TextField
    @Inject
    private String addToCartPlaceHolder;

    public String getAddToCartPlaceHolder() {
        return addToCartPlaceHolder;
    }

    @DialogField(fieldLabel = "Multiple Items Label",
            fieldDescription = "The text to display on the multiple items link",
            ranking = 3)
    @TextField
    @Inject
    private String multipleItemsLabel;

    public String getMultipleItemsLabel() {
        return multipleItemsLabel;
    }

    @DialogField(fieldLabel = "Multiple Items Link",
            fieldDescription = "The path of the multiple items link",
            required = true,
            ranking = 4)
    @PathField(rootPath = WatersConstants.ORDER_ROOT_PATH)
    @LinkInject
    private Link multipleItemsLink;

    public String getMultipleItemsLink() {
        return multipleItemsLink.getPath();
    }

    @DialogField(fieldLabel = "Multiple Items Icon",
            fieldDescription = "Select or enter the icon URL",
            ranking = 5)
    @PathField(rootPath = WatersConstants.DAM_PATH)
    @LinkInject(inherit = true)
    private Link multipleItemsIcon;

    public String getMultipleItemsIcon() { return multipleItemsIcon.getPath(); }

    @DialogField(fieldLabel = "Add Items Icon",
            fieldDescription = "Select or enter the icon URL",
            ranking = 6)
    @PathField(rootPath = WatersConstants.DAM_PATH)
    @LinkInject(inherit = true)
    private Link addItemsIcon;

    public String getAddItemsIcon() { return addItemsIcon.getPath(); }

    public String getAddToCartUrl() {
        return watersCommerceService.getAddToCartUrl();
    }

    @Nonnull
    @Override
    public String getExportedType() {
        return null;
    }
}
