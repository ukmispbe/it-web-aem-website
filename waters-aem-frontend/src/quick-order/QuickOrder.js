import React, { useRef, useState, useEffect, useCallback, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';
import ReactSVG from 'react-svg';

import AddToCart from '../sku-details/views/addToCart';
import AddToCartBody from '../sku-details/views/addToCartModal';
import Modal, { Header, keys } from '../utils/modal';
import Input from '../components/Input/Input';
import ScreenSizes from '../scripts/screenSizes';
import { mainCartContext } from "../analytics";

function QuickOrder(props) {
    const {
        buttonLabel,
        addToCartPlaceHolder,
        addToCartUrl,
        multipleItemsLabel,
        multipleItemsLink,
        addItemsIcon,
        multipleItemsIcon,
        isCommerceApiMigrated,
        showLabel,
        titleText,
        price,
        skuConfig
    } = props;
    const childRef = useRef();
    const [sku, setSku] = useState('');
    const [errorObjCart, setErrorObjCart] = useState({});
    const [modalShown, setModalShown] = useState(false);
    const [isMobile, setIsMobile] = useState(ScreenSizes.isMobile());
    const [modalConfig, setModalConfig] = useState({ ...skuConfig.modalInfo, partNumberLabel: skuConfig.skuInfo.partNumberLabel });

    function onChange(value) {
        setSku(value);
        childRef.current.onChangeSku(value);
    }

    // HideShow SKU Error
    function skuErrorMgs(showError, skuNumber = '') {
        try {
            const element = document.querySelector('.cmp-notification--error');
            if (element) {
                if (showError) {
                    document.querySelector('.cmp-notification--error .cmp-notification-description')
                        .innerText = `${skuNumber} ${document.querySelector('.cmp-notification--error .cmp-notification-title').innerText}`;
                    element.classList.remove('cmp-notification--dynamic');
                } else {
                    document.querySelector('.cmp-notification--error .cmp-notification-description').innerText = '';
                    element.classList.add('cmp-notification--dynamic');
                }
            }
        } catch (error) { }
    };

    // Hide SKU title msg on ComponentWillMount
    useLayoutEffect(() => {
        if (document.querySelector('.cmp-notification--error .cmp-notification-title')) {
            document.querySelector('.cmp-notification--error .cmp-notification-title').style.display = 'none';
        }
    }, []);

    // Call During Screen resize
    const updateViewport = useCallback(() => {
        setIsMobile(ScreenSizes.isMobile());
    }, [setIsMobile]);

    // Register updateViewport method on screen resize
    useEffect(() => {
        window.addEventListener('resize', updateViewport, true);
    }, [updateViewport]);

    // Get Added SKU info
    const skuResponse = useCallback(response => {
        if (response.cartModifications && response.cartModifications.length > 0) {
            if (response.cartModifications[0].entry && response.cartModifications[0].entry.product) {
                const { code, name } = response.cartModifications[0].entry.product;
                setModalConfig(currentVal => ({
                    ...currentVal,
                    textHeading: code,
                    text: name
                }));
            }
        }
    }, [setModalConfig, setErrorObjCart, setModalShown]);
    // Used for modal status
    const toggleModal = useCallback(() => {
        childRef.current.onChangeSku('');
        childRef.current.skuQuantityInput({ target: { value: 1 } });
        setSku(() => '');
        setErrorObjCart(() => ({}));
        skuErrorMgs(false);
        setModalShown(status => !status);
    }, [setSku, setErrorObjCart, skuErrorMgs, setModalShown, childRef]);
    // Error scenarios
    const toggleErrorModal = useCallback(error => {
        if (Object.keys(error).length > 0) {
            if ([400, 401, 404, 500].includes(error.status)) {
                setModalConfig(currentVal => ({
                    ...currentVal,
                    isOrderDetails: true,
                    textHeading: skuConfig.errorInfo.title,
                    text: skuConfig.errorInfo.wereSorry
                }));
                setErrorObjCart(error);
                setModalShown(true);
            } else {
                skuErrorMgs(true, childRef.current.state.skuNumber);
            }
        }
    }, [skuErrorMgs, setErrorObjCart, setModalConfig, setModalShown]);

    return (
        <>
            <div className="quick-order-parent">
                <Input
                    id="code"
                    type="text"
                    name="quickOrder"
                    placeholder={addToCartPlaceHolder}
                    value={sku}
                    className="cmp-sku-details__quantity quick-order-sku"
                    showLabel={showLabel}
                    onChange={onChange}
                    elementLocator="input-quick-order-sku"
                />
                <AddToCart
                    skuNumber={sku}
                    addToCartQty={1}
                    addToCartLabel={buttonLabel}
                    addToCartUrl={addToCartUrl}
                    isCommerceApiMigrated={isCommerceApiMigrated}
                    toggleParentModal={toggleModal}
                    toggleErrorModal={toggleErrorModal}
                    analyticsConfig={{ sku, price, context: mainCartContext, name: titleText }}
                    onRef={ref => { childRef.current = ref; }}
                    skuResponse={skuResponse}
                />
            </div>
            <a
                href={multipleItemsLink}
                className="quick-order-multiple-item"
                data-locator="link-quick-order-add-multiple-item"
            >
                <ReactSVG src={isMobile ? addItemsIcon : multipleItemsIcon} wrapper='span' />
                {multipleItemsLabel}
            </a>
            <Modal isOpen={modalShown} onClose={toggleModal} className='cmp-add-to-cart-modal'>
                {!(Object.keys(errorObjCart).length !== 0) && (
                    <Header
                        title={modalConfig.title}
                        icon={modalConfig.icon}
                        className={keys.HeaderWithAddedMarginTop} />
                )}
                {(Object.keys(errorObjCart).length !== 0) && (
                    <Header
                        title={skuConfig.errorInfo.title}
                        icon={skuConfig.errorInfo.icon}
                        className={keys.HeaderWithAddedMarginTopError}
                    />
                )}
                <AddToCartBody
                    config={modalConfig}
                    errorObjCart={errorObjCart}
                ></AddToCartBody>
            </Modal>
        </>
    )
}

QuickOrder.defaultProps = {
    buttonLabel: PropTypes.string,
    addToCartPlaceHolder: PropTypes.string,
    addToCartUrl: PropTypes.string,
    multipleItemsLabel: PropTypes.string,
    multipleItemsLink: PropTypes.string,
    addItemsIcon: PropTypes.string,
    multipleItemsIcon: PropTypes.string,
    showLabel: PropTypes.bool,
    isCommerceApiMigrated: PropTypes.bool,
    titleText: PropTypes.string,
    price: PropTypes.string,
    skuConfig: PropTypes.object,
}

QuickOrder.defaultProps = {
    buttonLabel: '',
    isCommerceApiMigrated: true,
    addToCartPlaceHolder: '',
    addToCartUrl: '',
    multipleItemsLabel: '',
    multipleItemsLink: '',
    addItemsIcon: '/content/dam/waters/en/brand-assets/icons/add.svg',
    multipleItemsIcon: '/content/dam/waters/en/brand-assets/icons/multiple.svg',
    showLabel: false,
    titleText: '',
    price: '',
    skuConfig: {},
}

export default QuickOrder;