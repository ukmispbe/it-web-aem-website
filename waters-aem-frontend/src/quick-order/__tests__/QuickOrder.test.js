import React from 'react';
import { shallow } from 'enzyme';

//Mocked Props
import QuickOrder from '../QuickOrder';
import props from '../QuickOrder.mock';
import AddToCart from '../../sku-details/views/addToCart';

describe('<QuickOrder />', () => {
    let enzymeWrapper;
    let addToCart;
    beforeAll(() => {
        Object.defineProperty(window, 'matchMedia', {
            value: () => ({
                matches: false,
                addListener: () => { },
                removeListener: () => { }
            }),
            writable: true,
            enumerable: true,
            configurable: true
        });
        enzymeWrapper = shallow(<QuickOrder {...props} />);
    });
    it('should render without throwing an error', () => {
        expect(enzymeWrapper).toMatchSnapshot();
    });
    it('should display default Add to cart section', () => {
        expect(enzymeWrapper.find('.quick-order-parent')).toHaveLength(1);
        expect(enzymeWrapper.find('#code')).toHaveLength(1);
        expect(enzymeWrapper.find(AddToCart)).toHaveLength(1);
        addToCart = enzymeWrapper.find(AddToCart).dive();
        expect(addToCart.find('.cmp-sku-details__quantity')).toHaveLength(1);
        expect(addToCart.find('.cmp-button')).toHaveLength(1);
        expect(enzymeWrapper.find('.quick-order-multiple-item')).toHaveLength(1);
    });
    it('should display Add Multiple Items', () => {
        expect(enzymeWrapper.find('.quick-order-multiple-item')).toHaveLength(1);
    });
    xit('should display ADD TO CART Button as active', () => {
        enzymeWrapper.find('#code').simulate('change', { target: { value: 'rertert' } });
        addToCart.find('.cmp-sku-details__quantity').simulate('change', { target: { value: '2' } });
    });
});