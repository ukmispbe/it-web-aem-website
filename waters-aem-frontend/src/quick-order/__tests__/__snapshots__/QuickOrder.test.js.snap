// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuickOrder /> should render without throwing an error 1`] = `ShallowWrapper {}`;
